PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader

```diff
@@ -36,48 +44,73 @@ in vec3 vNormal;
 
 layout(location = 0) out vec4 color;
 
-// calculates ambient light
-vec4 ambientlLight(DiffuseLight ambientLight) {
-	return vec4(ambientLight.Color.rgb * max(ambientLight.Intensity, 0.0), 1.0);
+vec3 ambientlLight(DiffuseLight ambientLight) {
+	return vec3(ambientLight.Color.rgb * max(ambientLight.Intensity, 0.0));
 }
 
-// calculates diffuse factor of any light
-vec4 directionalLight(DirectionalLight light, vec3 normalWS) {
+Phong directionalLight(DirectionalLight light, vec3 positionWS, vec3 normalWS) {
+	Phong result;
 	DiffuseLight base = light.Base;
 	vec3 dir = light.Direction.xyz;
 	float dirNdotL = max(dot(normalWS, dir), 0.0);
-	vec4 dirDiffuse = vec4(base.Color.rgb, 1.0) * max(base.Intensity, 0.0) * dirNdotL;
-	return dirDiffuse;
+	result.Diffuse = vec3(base.Color.rgb) * max(base.Intensity, 0.0) * dirNdotL;
+	
+	vec3 reflectDir = reflect(-dir, normalWS);
+	float spec = pow(max(dot(uCameraDirection.xyz, reflectDir), 0.0), 32);
+	result.Specular = uSpecularStrength * spec * base.Color.rgb;
```

Comment:
Specular color is not always equal to base color. Specular color should be taken from the material. Correct me if i'm wrong.

Reply:
Debugging it right now. I've spotted in that version wrongly added multiple lights influence resulting in grayish objects in bright scenes. Expect Blinn-Phong pt. 2 with fixes/improvements.